java/ql/src/semmle/code/java/frameworks/guava/Base.qll

```diff
@@ -0,0 +1,41 @@
+/** Definitions of flow steps through utility methods of `com.google.common.base`. */
+
+import java
+private import semmle.code.java.dataflow.ExternalFlow
+
+private class GuavaBaseCsv extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        //"package;type;overrides;name;signature;ext;inputspec;outputspec;kind",
+        "com.google.common.base;Strings;false;emptyToNull;(String);;Argument[0];ReturnValue;value",
+        "com.google.common.base;Strings;false;nullToEmpty;(String);;Argument[0];ReturnValue;value",
+        "com.google.common.base;Strings;false;padStart;(String,int,char);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Strings;false;padEnd;(String,int,char);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Strings;false;repeat;(String,int);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Strings;false;lenientFormat;(String,Object[]);;Argument;ReturnValue;taint",
+        "com.google.common.base;Joiner;false;on;(String);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;skipNulls;();;Argument[-1];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;useForNull;(String);;Argument[-1];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;useForNull;(String);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;withKeyValueSeparator;(String);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;withKeyValueSeparator;(String);;Argument[-1];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;withKeyValueSeparator;(char);;Argument[-1];ReturnValue;taint",
+        "com.google.common.base;Joiner;false;appendTo;;;Argument;ReturnValue;taint",
+        "com.google.common.base;Joiner;false;appendTo;;;Argument;Argument[0];taint",
+        "com.google.common.base;Joiner;false;appendTo;;;Argument[0];ReturnValue;value",
+        "com.google.common.base;Joiner;false;join;;;Argument;ReturnValue;taint",
+        "com.google.common.base;Joiner$MapJoiner;false;useForNull;(String);;Argument[0];ReturnValue;taint",
+        "com.google.common.base;Joiner$MapJoiner;false;useForNull;(String);;Argument[-1];ReturnValue;taint",
+        "com.google.common.base;Joiner$MapJoiner;false;appendTo;;;Argument;ReturnValue;taint",
+        "com.google.common.base;Joiner$MapJoiner;false;appendTo;;;Argument;Argument[0];taint",
+        "com.google.common.base;Joiner$MapJoiner;false;appendTo;;;Argument[0];ReturnValue;value",
```

Comment:
We should probably leave ourselves some hints about the signature when we get to convert these to use proper collection flow. A method like `StringBuilder appendTo(StringBuilder builder, Iterator<? extends Map.Entry<?,?>> entries)` will likely get an input spec along the lines of `Content[.value] of Content[Element] of Argument[1]`. I've also briefly talked to @smowton about simplifying the taint models a bit for methods that also have a value-preserving step, since the spec `"Argument[0];ReturnValue;value"` means that we ought to be able to infer `"Argument;ReturnValue;taint"` from `"Argument;Argument[0];taint"` and vice versa, such that only one of them is needed.

Reply:
Should I list the signatures (which there are quite a few of) in a comment, or directly in the CSV format? Or note them somewhere else?
Or just a comment that the signatures involve collection flow?

Reply:
When I remove either the `Argument;ReturnValue;taint` lines or the `Argument;Argument[0];taint` lines, my test cases fail.

Reply:
We need to get this commit https://github.com/github/codeql/pull/5383/commits/9b8e0fc9e8d4ea11146acc70ce6881b622a1ec4a merged in order to make one of those lines superfluous. I've asked Chris to put it into its own PR so we can get it merged.
As for leaving ourselves a comment, then it's probably sufficient to note the the signatures involve collection flow.

Reply:
This should work now that https://github.com/github/codeql/pull/5413 is merged.